xarray/core/weighted.py

```diff
@@ -0,0 +1,307 @@
+
+
+class Weighted:
+    """A object that implements weighted operations.
+
+    You should create a Weighted object by using the `DataArray.weighted` or
+    `Dataset.weighted` methods.
+
+    See Also
+    --------
+    Dataset.weighted
+    DataArray.weighted
+    """
+
+    __slots__ = ("obj", "weights")
+
+    @overload
+    def __init__(self, obj: "DataArray", weights: "DataArray") -> None:
+        ...
+
+    @overload
+    def __init__(self, obj: "Dataset", weights: "DataArray") -> None:
+        ...
+
+    def __init__(self, obj, weights) -> None:
+        """
+        Create a Weighted object
+
+        Parameters
+        ----------
+        obj : DataArray or Dataset
+            Object over which the weighted reduction operation is applied.
+        weights : DataArray
+            An array of weights associated with the values in this obj.
+            Each value in the obj contributes to the reduction operation
+            according to its associated weight.
+
+        Note
+        ----
+        Missing values in the weights are replaced with 0 (i.e. no weight).
+
+        """
+
+        from .dataarray import DataArray
+
+        msg = "'weights' must be a DataArray"
+        assert isinstance(weights, DataArray), msg
+
+        self.obj = obj
+        self.weights = weights.fillna(0)
+
+    def __repr__(self):
+        """provide a nice str repr of our Weighted object"""
+
+        msg = "{klass} with weights along dimensions: {weight_dims}"
+        return msg.format(
+            klass=self.__class__.__name__, weight_dims=", ".join(self.weights.dims)
+        )
+
+
+class DataArrayWeighted(Weighted):
+    def sum_of_weights(
+        self, dim: Optional[Union[Hashable, Iterable[Hashable]]] = None, axis=None
+    ) -> "DataArray":
+
+        return _sum_of_weights(self.obj, self.weights, dim=dim, axis=axis)
+
+    def sum(
+        self,
+        dim: Optional[Union[Hashable, Iterable[Hashable]]] = None,
+        axis=None,
+        skipna: Optional[bool] = None,
+        **kwargs
+    ) -> "DataArray":
+
+        return _weighted_sum(
+            self.obj, self.weights, dim=dim, axis=axis, skipna=skipna, **kwargs
+        )
+
+    def mean(
+        self,
+        dim: Optional[Union[Hashable, Iterable[Hashable]]] = None,
+        axis=None,
+        skipna: Optional[bool] = None,
+        **kwargs
+    ) -> "DataArray":
+
+        return _weighted_mean(
+            self.obj, self.weights, dim=dim, axis=axis, skipna=skipna, **kwargs
+        )
+
+
+# add docstrings
+DataArrayWeighted.sum_of_weights.__doc__ = _SUM_OF_WEIGHTS_DOCSTRING.format(
+    cls="DataArray"
+)
+DataArrayWeighted.mean.__doc__ = _WEIGHTED_REDUCE_DOCSTRING_TEMPLATE.format(
+    cls="DataArray", fcn="mean"
+)
+DataArrayWeighted.sum.__doc__ = _WEIGHTED_REDUCE_DOCSTRING_TEMPLATE.format(
+    cls="DataArray", fcn="sum"
+)
+
+
+class DatasetWeighted(Weighted):
+    def _dataset_implementation(self, func, **kwargs) -> "Dataset":
```

Comment:
Not for this PR, but we do this lots of places; we should find a common approach at some point!